src/DIRAC/Resources/IdProvider/IdProviderFactory.py

```diff
@@ -10,52 +10,112 @@
 from __future__ import division
 from __future__ import print_function
 
+import jwt as _jwt
+
 from DIRAC import S_OK, S_ERROR, gLogger
-from DIRAC.Core.Utilities.ObjectLoader import ObjectLoader
-from DIRAC.ConfigurationSystem.Client.Helpers.Resources import getInfoAboutProviders
+from DIRAC.Core.Utilities import ObjectLoader, ThreadSafe
+from DIRAC.Core.Utilities.DictCache import DictCache
+from DIRAC.ConfigurationSystem.Client.Helpers.Resources import getProviderInfo, getSettingsNamesForIdPIssuer
+from DIRAC.ConfigurationSystem.Client.Utilities import getAuthorisationServerMetadata
+from DIRAC.FrameworkSystem.private.authorization.utils.Clients import DEFAULT_CLIENTS
 
 __RCSID__ = "$Id$"
 
+gCacheMetadata = ThreadSafe.Synchronizer()
+
 
 class IdProviderFactory(object):
 
-  #############################################################################
   def __init__(self):
     """ Standard constructor
     """
     self.log = gLogger.getSubLogger('IdProviderFactory')
+    self.cacheMetadata = DictCache()
+
+  @gCacheMetadata
+  def getMetadata(self, idP):
+    return self.cacheMetadata.get(idP)
+
+  @gCacheMetadata
+  def addMetadata(self, idP, data, time=24 * 3600):
+    if data:
+      self.cacheMetadata.add(idP, time, data)
+
+  def getIdProviderForToken(self, token):
+    """ This method returns a IdProvider instance corresponding to the supplied
+        issuer in a token.
+
+        :param str token: token
+
+        :return: S_OK(IdProvider)/S_ERROR()
+    """
+    data = {}
+
+    # Read token without verification to get issuer
+    issuer = _jwt.decode(token, options=dict(verify_signature=False))['iss'].strip('/')
```

Comment:
How fast is decoding the token? Is it expensive enough that there should be a token class that caches the decoded result?

Reply:
I updated the public keys separately via IOLoop, now the token check takes 0.00245594978333.